airflow/utils/helpers.py

```diff
@@ -142,83 +135,6 @@ def as_flattened_list(iterable):
     return [e for i in iterable for e in i]
 
 
-def chain(*tasks):
```

Comment:
This is a very sad change because it is not backward compatible.  One release from the 1.10+ series tried to delete these functions. Users were very sad when this happened.  Maybe we should move all the other functions to a separate file, and leave these functions in the old place?

Reply:
I agree it's a bit sad (that's why I added UPDATING.md) but we already have a number of incompatibilities and we plan to have a migration tool to handle this https://issues.apache.org/jira/browse/AIRFLOW-6390 (and such automated conversion will be easy to add).

There are other functions that look useful - validate_key, reduce_in_chunks, as_tuple, as_flattened_list. I see them as being used by the users - they are not base-operator dependent but still useful either in custom operators or dags. 

I could of course move the other methods elsewhere, but I think there is an advantage of moving it (and any future similar methods) to BaseOperator. 

I wonder what others think about it ? Is it worth to introduce such incompatibility or not? 

I am happy to do both but I think it's quite a bit cleaner as it is now.

Reply:
The existence of the idea of creating an automatic tool has no effect on whether we should or should not make changes.

Others look much less used, although there may actually be someone who finds these functions. But in my opinion, this is the choice of the lesser evil. 

I personally don't see the benefits of creating static methods in Python. Why do you think it's worth creating a static method? It would be easier for me to write
```
chain ([taskA], [taskB])
```
than
```
BaseOperator.chain ([taskA], [taskB])
```
It is not possible to import a single static method into a file. You must always import the whole class. A similar opinion was expressed in "[Google Python Style Guide](http://google.github.io/styleguide/pyguide.html)":
> Use decorators judiciously when there is a clear advantage. Avoid @staticmethod and limit use of @classmethod.

Reply:
> The existence of the idea of creating an automatic tool does not affect whether we should or should not make changes.

I don't agree. By having automated tools we might decide that we deliberately not keep backwards compatibility in some cases more easily  - especially if those changes can easily and automatically be migrated in the existing DAGs. 

"Backwards compatibility" is simply not as important for 2.0 as it was in 1.10. This decision has already been made. We have a number of backwards incompatible changes, there is no going back.

> Others look much less used, although there may be someone who finds these functions. But in my opinion, this is the choice of the lesser evil.
I am not sure which one is lesser. If we automate migration for chain method it's not evil at all for migration. Especially that finally it makes sense for the chain method to be in baseoperator package (or in some form of operator_helpers but I prefer not to use helpers/managers/etc. if not neccessary). It makes much more sense for the other helpers method to be there - they were there originally, simply adding chains and the other method here later was a mistake that we want to correct now. For me more evil is to have bad naming in future version of airflow.

>> It is not possible to import a single static method into a file. You must always import the whole class. A similar opinion was expressed in "Google Python Style Guide":

Agree - will make it module function in baseoperator. it makes more sense. It comes from the old Java background where I prefer to have functions in classes.

Reply:
I definitely agree with @mik-laj  that it should not be a static method. Module Function is fine to me, the current change looks sensible @potiuk 👍

Reply:
Agree. Static was bad idea.